### PR TITLE
Add magic placeholder assets under lib, web, and test

### PR DESCRIPTION
**Pull request by jakemac53**
These assets are available to use as primary inputs, but cannot be read (and will give an `AssetNotFoundException` if you do try to read them). They are represented by a new type of node, `PlaceholderAssetNode`.

Note that this is a potentially breaking change for builders which have over-broad input extensions (looking at you `CopyBuilder`). These builders may attempt to read the input and throw an exception, but I don't think this is a major issue since it should be a very uncommon case. It may however break existing tests for people using `CopyBuilder` (it broke a bunch of ours).

I also added a bunch of boolean getters to describe functionality of the different asset node types, and replaced is checks where possible with those.

Closes https://github.com/dart-lang/build/issues/710

### CODE REVIEW

**Comment by natebosch** [comment]
This also addresses some of #689

File: build_runner/bin/create_merged_dir.dart [hunk 1]
```diff
@@ -72,7 +72,7 @@ Future main(List<String> args) async {
       assetGraph, packageGraph.root.name);
 
   for (var node in assetGraph.allNodes) {
-    if (node is SyntheticAssetNode) continue;
+    if (!node.isReadable) continue;
     if (node is GeneratedAssetNode && !node.wasOutput) continue;
```
Comment (natebosch) [inline build_runner/bin/create_merged_dir.dart]: I'm curious why this isn't covered by `isReadable`.

File: build_runner/lib/src/asset_graph/node.dart [hunk 2]
```diff
@@ -27,8 +27,34 @@ abstract class AssetNode {
   /// exist.
   Digest lastKnownDigest;
 
+  /// Whether or not this node was an output of this build.
+  bool get isGenerated => false;
+
+  /// Whether or not this asset can be read.
```
Comment (natebosch) [inline build_runner/lib/src/asset_graph/node.dart]: "Whether or not this asset type can be read.

This does not indicate whether the node is readable at this moment in time."

File: build_runner/lib/src/asset_graph/node.dart [hunk 3]
```diff
@@ -27,8 +27,34 @@ abstract class AssetNode {
...
+  bool get isReadable => true;
+
+  /// Whether or not this node can be used as a primary or secondary input.
```
Comment (natebosch) [inline build_runner/lib/src/asset_graph/node.dart]: Expand this, mention that some internal assets can be read but can't be used as primary inputs.

Comment (natebosch) [inline build_runner/lib/src/asset_graph/node.dart]: what is a secondary input?

**Comment by jakemac53** [comment]
working on changelog updates as well now

File: build_runner/bin/create_merged_dir.dart (hunk 1 above)
Comment (natebosch) [reply build_runner/bin/create_merged_dir.dart]: Ah we discussed this and I forgot to delete the comment. Ignore :)

File: build_runner/lib/src/asset_graph/node.dart (hunk 2 above)
Comment (jakemac53) [reply build_runner/lib/src/asset_graph/node.dart]: done

File: build_runner/lib/src/asset_graph/node.dart (hunk 3 above)
Comment (jakemac53) [reply build_runner/lib/src/asset_graph/node.dart]: done

Comment (jakemac53) [reply build_runner/lib/src/asset_graph/node.dart]: any input that isn't primary (something you accessed through `buildStep`)

Comment (jakemac53) [reply build_runner/lib/src/asset_graph/node.dart]: updated this to only mention primary inputs as that was actually the intention behind it

**Pull request merged**